### PR TITLE
MM-29979: make websocket writes zero-alloc

### PR DESCRIPTION
**Pull request by agnivade**
This reverts commit 31dac80e66c31a123e56b116e8173f64d1a88d2d.

### CODE REVIEW

**Comment by agnivade** [comment]
Well, we are back to do a bit of allocations. So it's not exactly "zero-alloc" anymore. But the precomputeJSON already allocates small byte buffers. So unless we remove that, there's no escaping this. So might as well accept the compromise.

File: app/web_conn.go [hunk 1]
```diff
@@ -201,20 +208,25 @@ func (wc *WebConn) writePump() {
...
 			}
 
-			var msgBytes []byte
+			buf.Reset()
+			var err error
 			if evtOk {
 				cpyEvt := evt.SetSequence(wc.Sequence)
-				msgBytes = []byte(cpyEvt.ToJson())
+				err = cpyEvt.Encode(enc)
 				wc.Sequence++
```
Comment (streamer45) [inline app/web_conn.go:216]: Unrelated to the PR and probably a minor thing but should the sequence get bumped also in case of error?

**Review by streamer45 — APPROVED** [review]
Looks good :+1:

File: app/web_conn.go (hunk 1 above)
Comment (agnivade) [reply app/web_conn.go:216]: Hmm.. good point. I think it should not. But I need to verify a bit how the sequence number is used by the client. In any case, let's visit this in a separate PR.

**Review by isacikgoz — APPROVED** [review]
LGTM 👍

**Comment by agnivade** [comment]
@lindalumitchell - This caused an issue earlier due to an oversight. Please do a quick smoke testing to ensure that all basic functionalities work, giving special focus to websocket events.

**Review by lindalumitchell — APPROVED** [review]
LGTM!

**Pull request merged**